File: lesson_14_20241119/hw/FindElement.java
package lesson_14_20241119.hw;

import java.util.HashMap;

public class FindElement {
    public static void main(String[] args) {
        int[] arr = {1, 7, 4, 3, 4, 8, 7};
        int n = arr.length;
        int k = 2;
        System.out.println(firstElementNative(arr, n, k));

        System.out.println(firstElementMap(arr, n, k));

    }
    public static int firstElementNative(int[] arr, int n, int k) {
        // This loop is used for selection of elements
        for (int i = 0; i < n; i++) {
            int count = 0; // Count how many times selected element occurs
            for (int j = i; j < n; j++) {
                if (arr[i] == arr[j]) {
                    count++;
                }
            }
            if (count == k) { // Check, if it occurs k times or not
                return arr[i];
            }
        }
        return -1;
    }

    public static int firstElementMap(int[] arr, int n, int k) {
        // unordered map to count occurrences of each element
        HashMap<Integer, Integer> map = new HashMap<>();
        for (int i = 0; i < n; i++) {
            int a = 0;
            if (map.get(arr[i]) != null) {
                a = map.get(arr[i]);
            }
            map.put(arr[i], a + 1);
        }
        for (int i = 0; i < n; i++) { // if count of element == k ,then it is the required first element
            if (map.get(arr[i]) == k) {
                return arr[i];
            }
        }
        return -1;
    }

   
}

File: lesson_14_20241119/hw/Find_element.js
function firstElementNative(arr, n, k) {
    // This loop is used for selection
    // of elements
    for (let i = 0; i < n; i++) {
        // Count how many times selected element
        // occurs
        let count = 0;
        for (let j = i; j < n; j++) {
            if (arr[i] === arr[j]) {
                count++;
            }
        }

        // Check, if it occurs k times or not
        if (count === k) {
            return arr[i];
        }
    }

    return -1;
}

// Эффективный подход через Map
// use this function:
function firstElementMap(arr, n, k) {
    let map = new Map();
    for (let i = 0; i < n; i++) {
        var a = 0;
        if(map.has(arr[i])) {
            a = map.get(arr[i]) || 0;
        }
        map.set(arr[i], a + 1);
    }
    for (let i = 0; i < n; i++) {
        if (map.get(arr[i]) === k) {
            return arr[i];
        }
    }
    return -1;
}


let arr = [1, 7, 4, 3, 4, 8, 7];
    let n = arr.length;
    let k = 2;
    console.log(firstElementNative(arr, n, k));
    console.log(firstElementMap(arr, n, k));